File: solutions/c/kindergarten-garden/1/kindergarten_garden.c
#include "kindergarten_garden.h"

#include <string.h>


plant_t get_plant_name(char plant) {
    switch (plant) {
        case 'G':
            return GRASS;
        case 'C':
            return CLOVER;
        case 'R':
            return RADISHES;
        case 'V':
            return VIOLETS;
    }
    return -1;
}


plants_t plants(const char *diagram, const char *student) {
    /* calculate the index of the student in the alphabet   */
    int index = student[0] - 'A';
    /* calculate the index of the plants in the diagram */
    int plants_index = index * 2;
    /* Find the index of \n in the diagram */
    char *newline_index = strchr(diagram, '\n');
    int row_two_index = (int)(newline_index - diagram) + 1 + plants_index;

    plants_t plants = { .plants = { get_plant_name(diagram[plants_index]), get_plant_name(diagram[plants_index + 1]), get_plant_name(diagram[row_two_index]), get_plant_name(diagram[row_two_index + 1]) } };
    return plants;
}

File: solutions/typescript/kindergarten-garden/1/kindergarten-garden.ts
//
// This is only a SKELETON file for the 'Kindergarten Garden' exercise.
// It's been provided as a convenience to get you started writing code faster.
//

const DEFAULT_STUDENTS: Student[] = [
  'Alice',
  'Bob',
  'Charlie',
  'David',
  'Eve',
  'Fred',
  'Ginny',
  'Harriet',
  'Ileana',
  'Joseph',
  'Kincaid',
  'Larry',
]

const PLANT_CODES = {
  G: 'grass',
  V: 'violets',
  R: 'radishes',
  C: 'clover',
} as const

type Student = string
type Plant = (typeof PLANT_CODES)[keyof typeof PLANT_CODES]
type Plants = Plant[]
type Pots = Plants[]

export class Garden {
  constructor(private diagram: string, private students = DEFAULT_STUDENTS) {
    this.students.sort()
  }

  public plants(student: Student): Plants {
    const studentIndex = this.students.indexOf(student)
    const pots = this.diagram.split('\n').map((row) => row.split(''))
    const studentPots = pots.map((row) => row.slice(studentIndex * 2, studentIndex * 2 + 2))
    const studentPlants = studentPots.flatMap((pot) => pot.map((plant) => PLANT_CODES[plant as keyof typeof PLANT_CODES]))
    return studentPlants as Plants
  }
}
